### ISSUE TITLE
GzipDecoder assume the input contains only one member

### ISSUE THREAD

**Issue by cpfs-clustertech** [shazow/urllib3#1285]
gzip files have a property that if you concatenate two gzip files, the result is still considered a gzip file:
```
$ echo hello > a
$ echo world > b
$ gzip a
$ gzip b
$ cat a.gz b.gz > c.gz
$ zcat c.gz
hello
world
```
Since they are still considered a correct gzip file, we should expect such files of "multiple members" to be returned by web requests.

When you use zlib to decompress the result, you need special precaution:
```
$ cat pyunzip
#!/usr/bin/env python
import zlib

obj = zlib.decompressobj(16 + zlib.MAX_WBITS)
with open('c.gz') as fin:
    while True:
        data = fin.read(1024 * 1024)
        if data == '':
            break
        inlen = len(data)
        data = obj.decompress(data)
        print inlen, len(data)
$ ./pyunzip
56 6
```
The result is 6 bytes, i.e., only contains the "hello" part, without the "world" part.  When a web request returns such a file, you end up having a "truncated" file as output: the output contains only the first member of the gzip file. You can detect this by the "unused_data" member of the decompress object:
```
$ cat pyunzip
#!/usr/bin/env python
import zlib

obj = zlib.decompressobj(16 + zlib.MAX_WBITS)
with open('c.gz') as fin:
    while True:
        data = fin.read(1024 * 1024)
        if data == '':
            break
        inlen = len(data)
        data = obj.decompress(data)
        print inlen, len(data), len(obj.unused_data)
$ ./pyunzip
56 6 28
```
We need to detect the case in the code in order to get the file fully:
```
$ cat pyunzip
#!/usr/bin/env python
import zlib

obj = zlib.decompressobj(16 + zlib.MAX_WBITS)
with open('c.gz') as fin:
    while True:
        if obj.unused_data:
            obj, data = zlib.decompressobj(16 + zlib.MAX_WBITS), obj.unused_data
        else:
            data = fin.read(1024 * 1024)
        if data == '':
            break
        inlen = len(data)
        data = obj.decompress(data)
        print inlen, len(data), len(obj.unused_data)
$ ./pyunzip
56 6 28
28 6 0
```
So we need similar code in `response.py` of urllib3 to support such files.  I tried the following which seemed to work (Python 2.7):
```
    def decompress(self, data):
        if not data:
            return data
        ret = self._obj.decompress(data)
        if self._obj.unused_data:
            self._obj, data \
                = zlib.decompressobj(16 + zlib.MAX_WBITS), self._obj.unused_data
            ret += self._obj.decompress(data)
        return ret
```

**Comment by haikuginger** [comment]
@cpfs-clustertech, thanks for bringing this up! We'd welcome a PR to handle this edge case in combination with the appropriate regression tests.

**Comment by sigmavirus24** [comment]
I absolutely agree with @haikuginger. This is something I wasn't aware of, and definitely not something I'd have known to check for otherwise. Thanks for writing this up in a comprehensible way with a minimal reproducible example to toy with. 💯

**Comment by cpfs-clustertech** [comment]
After a second thought, the second "if" in my fix really needs to be a "while".

**Comment by isaacto** [comment]
See https://github.com/shazow/urllib3/pull/1289.

**Comment by sigmavirus24** [comment]
Closed by #1289
